linkerd/app/inbound/src/lib.rs

```diff
@@ -128,11 +146,16 @@ impl Inbound<()> {
         gateway: G,
     ) -> (SocketAddr, impl Future<Output = ()> + Send)
     where
+        B: Bind + Clone + Send + Sync + 'static,
+        B::Addrs: Param<Remote<ClientAddr>>
+            + Param<Local<ServerAddr>>
+            + Param<OrigDstAddr>
+            + Clone
+            + Send,
+        B::Io: io::Peek + io::PeerAddr + Debug + Unpin,
```

Comment:
we could avoid having to write these bounds out in as many places by just making them into bounds on the definition of the `Bind` trait, since it turns out everything requires these traits as well. I wasn't sure if it was worth making the trait definition more restrictive or not...

Reply:
also, we *could* try to figure out a way to get rid of the `PeerAddr` trait and just get client addresses from the target always. IDK if this matters or is worth doing...

Reply:
we need peeraddr for now at least, as the http server marks extensions with the client addr, but we can't have the http server created per-client (since we may need to do cached resolutions before this) -- so we'd either need to rework caching or have an Io type that supports extensions...

Reply:
yeah, i took a look at that on Monday and i think that we're better off keeping it, and if we are going to get rid of it, that's out of scope for this PR.

Reply:
@olix0r 7b16586 moves all the bounds on the `Bind::Io` type to the `Bind` trait so they're not repeated everywhere else (and removes some of the `Bind::Addrs` bounds that the `Bind` trait already requires).

We _could_ also put the `Param<Local<ServerAddr>>` and `Param<Remote<ClientAddr>>` bounds on `Bind::Addrs` so they aren't repeated, but it felt a _little_ weird to see it require only `Param<OrigDstAddr>` at use-sites; i could go either way on this...